demo/sizzle.sysl

```diff
@@ -0,0 +1,318 @@
+        return ok
+
+    !type Customer:
+        @description="Empty Type"
+        first_name <: string
+        middle_name <: string
+        last_name <: string
+        phone <: string
+        email <: string
+
+    !type NewCustomerRequest:
+        first_name <: string
+        middle_name <: string
+        last_name <: string
+        phone <: string
+        email <: string
+        password <: string
+
+    !type UnauthorizedError:
+        error_msg <: string
+
+ProductServer:
+    @package="BFF"
+    @description =: 
+        | This server handles all the product
+        | related endpoints.
+
+    Menu:
+        # TODO: DB/cache op
+        return ok <: Products
+
+    !alias Products:
+        sequence of Product
+
+    !type Product:
+        @description="Product information"
+        id <: int
+        name <: string
+        image <: string
+        details <: string
+        price <: int
+
+
+OrderServer:
+    @package="BFF"
+    @description =: 
+        | This server handles all the order
+        | related endpoints.
+
+    Order(req <: OrderRequest):
+        if order_id is nil:
+            # TODO: DB op create a new order
+            return ok <: Order
+        else:
+            # TODO: add/update OrderProduct for the 
+            # existing order
+            return ok <: Order
+
+    UpdateOrderStatus(order_id <: int, status <: int):
+        # TODO: DB op get order by order_id
+        # state machine
+        return ok <: Order
+
+    Review(score <: int, comment <: string):
+        # TODO: DB op
+        return ok <: Order
+
+    !type OrderRequest:
+        order_id <: int
+        product_id <: int
+        quantity <: int
+
+    !type Order:
+        @description="Customer order information"
+        id <: int
+        status <: OrderStatus
+        total_price <: int
+        items <: sequence of OrderProduct
+        paid <: bool
+        review_score <: int
+        review_comment <: string
+
+    !type OrderProduct:
+        @description="Order items"
+        product_id <: int
+        quantity <: int
+        comments <: string
+
+    !enum OrderStatus:
+        created: 1
+        placed: 2
+        shipped: 3
+        delivered: 4
+    
+
+PaymentServer:
+    @package="BFF"
+    @description =: 
+        | This server handles all the payment related endpoints.
+
+    Pay:
+        if processor_type == "visa":
+            Visa <- Pay
+        else if processor_type == "mastercard":
+            MasterCard <- POST /pay
+        else:
+            return 500 < NotSupportedError
+
+        return 200
+
+    !enum PaymentType:
+        cash: 1
+        card: 2
+
+Visa[~external]:
+    # ~external specifies an external service
+    Pay:
+        ...
+
+DeliveryServer:
+    @package="BFF"
+    @description =: 
+        | We are going to provide delivery service ASAP
+        | since our customers need it during COVID-19
+    ...
+
+
+
+MegaDatabase[~db]:
+    # ~db specifies a database
```

Comment:
This generates a really nice type diagram!

![image](https://user-images.githubusercontent.com/25007585/88146342-1543eb00-cc3f-11ea-8349-aec8e83ad95b.png)

Reply:
Yeah I like it

Reply:
It looks great! The data model itself is a bit odd though:

- Each review should probably have a customer as an author, not be related to an order.
- I think card_payment is just a card?
- All names should be consistently singular or plural
- What is orders_products.comments?
- payment_details.type is a float?

Nits:
- Probably get rid of customers.middle_name?
- Maybe add some dates to things - dates are always a pain

Reply:
orders_products.comments is something like "less chilli", "more sugar", etc.